scripts/measures/2018/import-quality-measures.js

```diff
@@ -10,207 +10,209 @@ const Constants = require('../../../constants.js');
- * @type {Object}
- *
- *  * `constant_fields` are fields which are the same for all measures being
- *  created from the CSV input.
- *  * `source_fields` are fields which should find values in the CSV input.
- *
  * NOTE: We currently don't differentiate between empty CSV fields and CSV fields
  * explicitly marked as false, because there are no optional fields with
  * true/false values. If there ever are, the default: structure / true/false_markers
  * below and the mapInput function would need to be updated.
  */
-const QUALITY_CSV_CONFIG = {
-  // markers are what the CSV creators chose as field values;
-  // they use different conventions for different columns
-  truthy_markers: ['true', 'x'],
-  falsy_markers: ['false', 'null', 'n/a'],
-  constant_fields: {
-    category: 'quality',
-    isRegistryMeasure: false,
-    isRiskAdjusted: false
-  },
-  sourced_fields: {
-    // fields are csv columns indexed starting from 0
-    title: 0,
-    eMeasureId: {
-      index: 1,
-      default: null
-    },
-    nqfEMeasureId: {
-      index: 2,
-      default: null
-    },
-    nqfId: {
-      index: 3,
-      default: null
-    },
-    measureId: 4,
-    description: 5,
-    nationalQualityStrategyDomain: 6,
-    measureType: {
-      index: 7,
-      mappings: { // there should be no capital letters in the keys below
-        'process': 'process',
-        'outcome': 'outcome',
-        'patient engagement/experience': 'patientEngagementExperience',
-        'efficiency': 'efficiency',
-        'intermediate outcome': 'intermediateOutcome',
-        'structure': 'structure',
-        'patient reported outcome': 'outcome',
-        'composite': 'outcome',
-        'cost/resource use': 'efficiency',
-        'clinical process effectiveness': 'process'
-      }
-    },
-    primarySteward: 8,
-    metricType: 50,
-    firstPerformanceYear: {
-      index: 51,
-      default: 2017
-    },
-    lastPerformanceYear: {
-      index: 52,
-      default: null
-    },
-    isHighPriority: {
-      index: 54,
-      default: false
-    },
-    isInverse: {
-      index: 55,
-      default: false
-    },
-    overallAlgorithm: 59,
-    isIcdImpacted: 64,
-    isToppedOutByProgram: 65
-  }
+
+// Constant fields are not present in the source CSV. They are
+// constant across quality measures, so we insert these fields into every measure
+const CONSTANT_FIELDS = {
+  category: 'quality',
+  isRegistryMeasure: false,
+  isRiskAdjusted: false
 };
 
-// mapping from quality measures csv column numbers to submission method array indices
+// Ignored fields are present in the source CSV but not imported
+// into measures data by this script
+const IGNORED_FIELDS = [
+  'eMeasureUuid',
+  'numStrataClaims',
+  'numStrataRegistry',
+  'numStrataEcqm'
+];
+
+// Main set of fields below mapped to their default values
+// Except for measure type which is a custom mapping
+// Undefined means the field has no default; no field should ever end up undefined
```

Comment:
One more question about the `undefined` vs `null` thing -- I could be wrong but I think there's a possibility that a field ends up `undefined`? Because line 266 is:
```
if (_.has(MAIN_FIELDS, fieldName)) {
```
and `_.has({someField: undefined}, 'someField')` does return true -- so if the `input` is undefined as well, that field will get set to `undefined`?